woocommerce/woocommerce
# Introduce a `reserved stock` class and database table to prevent race conditions during checkout

### All Submissions:


### Changes proposed in this Pull Request:

This PR is a result of the work in https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1652 and based on @claudiosanches PR (https://github.com/woocommerce/woocommerce/pull/25172) which was in turn based on the work by @vedanshujain (https://github.com/woocommerce/woocommerce/pull/25089) :)

The `ReserveStock` class is from the blocks plugin, and it's been updated with the atomic DB query to ensure stock is only reserved if available. I will be following up with an update on Blocks to match (this will be used during transition to the core class so we can use it before `4.0`).

The original query used meta:

```
INSERT INTO {order_table} VALUES (order_id, meta_key, quantity)
SELECT {order_id}, {meta_key}, {quantity} FROM DUAL 
WHERE {query_for_stock} - {query_for_held_stock} >= {required_stock}
```

This version instead uses a custom DB table, and stores the order ID which along with the product ID of the line item makes up the primary key to prevent duplicates.

```
REPLACE INTO {$wpdb->wc_reserved_stock} ( order_id, product_id, stock_quantity, expires )
SELECT %d, %d, %d, ( NOW() + INTERVAL %d MINUTE ) from DUAL
WHERE ( $query_for_stock FOR UPDATE ) - ( $query_for_reserved_stock FOR UPDATE ) >= %d
```

If this fails for any item in the order, an Exception will be thrown which will prevent checkout. 

Rather than call the new class (and helper functions) directly, I've added some additional hooks so all stock reservation logic can be kept together.

The main sticking point I think will be the `ReserveStock` class - I namespaced it in the `src` directory but I don't know if that's where we want to place it. Either way, the use of a namespace will prevent it conflicting with the version in Blocks.

I've updated the tests from the other PR to reflect this class not being part of the order class -- there are tests in place to check the numbers of held stock after creating orders. Tests are passing locally.

Closes #24009

### How to test the changes in this Pull Request:

I tested this following the instructions @claudiosanches gave to ensure it covered the concurrency issue. Pasting from the other PR:

> Unfortunately, this requires simulating race conditions so this is not trivial to test. I'd recommend using this plugin https://github.com/Automattic/wc-load-testing/ to simulate race conditions on a test site and using a config where stock quantity is less than the number of request like so:

1. Install the plugin by following the instructions on the README page of the plugin. Activate the plugin.
2. Go to Tools > WooCommerce Load Testing in the wp-admin page on your site.
3. Scroll down to the Process Checkout test section.
4. Use a config so that number of stocks are less then number of requests like so:

<img width="997" alt="69406090-3a3a3400-0d27-11ea-9d67-07578df4685c" src="https://user-images.githubusercontent.com/90977/74844067-ceb33300-5324-11ea-91aa-193329c9b24b.png">

Run the test, then go the Products page. The most recent product will have negative stock without this patch. Apply this patch and run test again, now the most recent product should have 0 stock.

(Make sure that cart is empty before running the test)

### Other information:




### Changelog entry

> Updated stock handling to prevent race conditions when orders come in at the same time.
